README.md

# odroid-fan-controller
A command line fan speed controller for ODROID XU3/XU4 models.

# Usage
Use `python odroid-fan-controller.py -m 0` to set the fan to: `manual` mode

Use `python odroid-fan-controller.py -s 100` to set the fan to: `100%`

Set the fan speed from 25%-100%. Below 25% the fan starts making a weird noise.

# How it works
ODROID controls the fan with a couple of special files in `/sys/devices/odroid_fan.13/`

    `/sys/devices/odroid_fan.13/pwm_duty` takes an integer 0 - 255. 255 being the 100% power.

    `/sys/devices/odroid_fan.13/fan_mode` specifies manual or automatic mode.

# Manual fan control without odroid-fan-controller
`echo 0 > /sys/devices/odroid_fan.13/fan_mode` turns the fan to manual mode.

`echo 255 > /sys/devices/odroid_fan.13/pwm_duty` gives the fan 100% power.

`echo 1 > /sys/devices/odroid_fan.13/fan_mode` turns the fan back to auto mode.



HISTORY.rst
=======
History
=======

0.1.0 (2019-04-22)
------------------

* First release on PyPI.


tests/test_odroid_fan_controller.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""Tests for `odroid_fan_controller` package."""

import pytest

from click.testing import CliRunner

from odroid_fan_controller import odroid_fan_controller
from odroid_fan_controller import cli


@pytest.fixture
def response():
    """Sample pytest fixture.

    See more at: http://doc.pytest.org/en/latest/fixture.html
    """
    # import requests
    # return requests.get('https://github.com/audreyr/cookiecutter-pypackage')


def test_content(response):
    """Sample pytest test function with the pytest fixture as an argument."""
    # from bs4 import BeautifulSoup
    # assert 'GitHub' in BeautifulSoup(response.content).title.string


def test_command_line_interface():
    """Test the CLI."""
    runner = CliRunner()
    result = runner.invoke(cli.main)
    assert result.exit_code == 0
    assert 'odroid_fan_controller.cli.main' in result.output
    help_result = runner.invoke(cli.main, ['--help'])
    assert help_result.exit_code == 0
    assert '--help  Show this message and exit.' in help_result.output


odroid_fan_controller/odroid_fan_controller.py
# !/usr/bin/python
import sys, getopt, os

from .constants import *

usage = "odroid-fan-controller.py [-m,--mode <0,1>,<'auto','manual'>] [-s, --speed <25-100>] "
usage += "[-v]"


def main(argv):
    get_fan_mode()
    get_fan_mode()

    try:
        opts, args = getopt.getopt(argv, "hm:s:p:v", ["mode=", "speed="])
    except getopt.GetoptError:
        sys.exit(2)
    for opt, arg in opts:
        if opt in ("-h", "--help"):
            print(usage)
            sys.exit()
        elif opt in ("-s", "--speed"):
            set_fan_speed(float(arg))
        elif opt in ("-m", "--mode"):
            set_fan_mode(arg)
        elif opt == "-v":
            status()


def status():
    print("Fan Speed: {0}%".format(get_fan_mode()))
    print("Fan Mode: {0}".format(get_fan_mode(str=True)))


def set_fan_speed(speed):
    # check fan mode
    if get_fan_mode() == 'manual':
        if speed < 25:
            print('Fan power below 25% causes a malfunction with the ODROID fan. Cancelling.')
            return

        speed = speed * 2.55

        write_to_file(odroid_pwm_duty_file, speed)
    elif get_fan_mode() == 'auto':
        print('Fan Mode set to "auto".\r\nPlease set to "manual" and try again.')


def set_fan_mode(fanmode):
    _mode = fanmode
    if (_mode == 'auto'):
        _mode = '1'
    elif (_mode == 'manual'):
        _mode = '0'
    print('_mode: {0}'.format(_mode))
    if (_mode == '1') or (_mode == '0'):
        print('get_fan_mode(str=False): {0}'.format(get_fan_mode()))
        if get_fan_mode(str=False) == _mode:
            print('Fan mode already set to {0}'.format(fanmode))
            return
        write_to_file(odroid_fan_mode_file, _mode)
        print('_mode: {0} written to {1}'.format(_mode, odroid_fan_mode_file))
        print('get_fan_mode(str=False)(after write): {0}'.format(get_fan_mode(str=False)))
    else:
        print('Invalid mode: {0}'.format(_mode))


def get_fan_mode(str=True):
    def number(mode):
        return '1' if 'auto' in mode else 'manual'

    mode = read_file(odroid_fan_mode_file)
    odroid_fan_mode = 'manual' if 'manual' in mode else 'auto'
    if str:
        return odroid_fan_mode
    else:
        return number(odroid_fan_mode)


def get_fan_speed():
    odroid_fan_speed = (float(read_file(odroid_pwm_duty_file)) / 2.55)
    if int(odroid_fan_speed) == 0:
        odroid_fan_speed = 0
    return odroid_fan_speed


def write_to_file(filePath, data):
    filePath = os.path.join(odroid_device_path, filePath)
    with open(filePath, 'w') as dataFile:
        dataFile.write(str(data))


def read_file(filePath):
    filePath = os.path.join(odroid_device_path, os.path.join(odroid_device_path, filePath))
    data = ''
    with open(filePath, 'r') as dataFile:
        data = dataFile.read()
        return data


if __name__ == '__main__':
    main(sys.argv[1:])


odroid-fan-controller.py
# !/usr/bin/python
import sys, getopt, io, os
usage = "odroid-fan-controller.py [-m,--mode <0,1>,<'auto','manual'>] [-s, --speed <25-100>] "
usage += "[-v]"
# odroid-xu4 device paths
odroid_device_path = '/sys/devices/odroid_fan.13/'
odroid_pwm_duty_file = 'pwm_duty'
odroid_fan_mode_file = 'fan_mode'
# odroid_fan_file_mode = 0 is manual fan mode
# odroid_fan_file_mode = 1 is auto fan mode
odroid_fan_file_mode = 0
odroid_fan_mode = ['manual', 'auto']
# odroid_pwm_duty_file: 0 is off, 255 is 100% fan power
# fan speed = [0 - 100]% * 2.55
odroid_fan_speed = 0

def main(argv):
    getfanspeed()
    getfanmode()
    try:
       opts, args = getopt.getopt(argv,"hm:s:p:v",["mode=","speed="])
    except getopt.GetoptError:
       print usage
       sys.exit(2)
    for opt, arg in opts:
      if opt in ("-h", "--help"):
          print usage
          sys.exit()
      elif opt in ("-s", "--speed"):
          setfanspeed(float(arg))
      elif opt in ("-m", "--mode"):
          setfanmode(arg)
      elif opt == "-v":
          status()
def status():
    print "Fan Speed: {0}%".format(getfanspeed())
    print "Fan Mode: {0}".format(getfanmode(str=True))
def setfanspeed(speed):
    # check fan mode
    if getfanmode(str=True) == 'manual':
        if speed < 25:
            print 'Fan power below 25% causes a malfunction with the odroid fan. Returning.'
            return
        odroid_fan_speed = speed
        speed = speed * 2.55
        writetofile(odroid_pwm_duty_file, speed)
    elif getfanmode() == 'auto':
        print 'Fan Mode set to "auto".\r\nPlease set to "manual" and try again.'
def setfanmode(fanmode):
    _mode = fanmode
    if(_mode == 'auto'):
        _mode = '1'
    elif(_mode == 'manual'):
        _mode = '0'
    print '_mode: {0}'.format(_mode)
    if (_mode == '1') or (_mode == '0'):
        print 'getfanmode(str=False): {0}'.format(getfanmode(str=False))
        if getfanmode(str=False) == _mode:
            print 'Fan mode already set to {0}'.format(fanmode)
            return
        writetofile(odroid_fan_mode_file, _mode)
        print '_mode: {0} written to {1}'.format(_mode, odroid_fan_mode_file)
        print 'getfanmode(str=False)(after write): {0}'.format(getfanmode(str=False))
    else:
        print 'Invalid mode: {0}'.format(mode)
def getfanmode(str=True):

    def number(mode):
        return '1' if 'auto' in mode else 'manual'

    mode = readfile(odroid_fan_mode_file)
    odroid_fan_mode = 'manual' if 'manual' in mode else 'auto'
    if str:
        return odroid_fan_mode
    else:
        return number(odroid_fan_mode)


def getfanspeed():
    odroid_fan_speed = (float(readfile(odroid_pwm_duty_file)) / 2.55)
    if int(odroid_fan_speed) == 0:
        odroid_fan_speed = 0
    return odroid_fan_speed

def writetofile(filePath, data):
    filePath = os.path.join(odroid_device_path, filePath)
    with open(filePath, 'w') as dataFile:
        dataFile.write(str(data))
def readfile(filePath):
    filePath = os.path.join(odroid_device_path, os.path.join(odroid_device_path, filePath))
    data = ''
    with open(filePath, 'r') as dataFile:
        data = dataFile.read()
        return data

if __name__ == '__main__':
    main(sys.argv[1:])
